driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
>  Context.Deadline() may be earlier than those two timeouts. So Context.Deadline() must be passed into these two functions. 

How about calling `.Close()` instead of `.SetDeadline()` when `<- ctx.Done()`?
Cancelled connection is mostly unusable furthermore.